[flang] Implement SIMPLE procedure support (parsing, semantics, diagnostics)

mlir-maiden opened the pull request:
Implement support for the Fortran 2023 `SIMPLE` procedure specifier.

This change includes:
- Parsing and attribute registration for `SIMPLE` procedures
- Adding `SIMPLE` to procedure characteristics
- Semantic support for `SIMPLE`, including treating `SIMPLE` procedures as `PURE` (`SIMPLE` implies `PURE`)
- Rejecting `SIMPLE` + `IMPURE` combinations
- Emitting and reading `SIMPLE` in mod files
- Adding parser, semantic, and modfile tests to validate `SIMPLE` behavior

`SIMPLE` procedures satisfy `PURE` requirements, while `PURE` procedures do not imply `SIMPLE`.

CC @klausler @eugeneepshteyn @akuhlens

mlir-maiden committed:
[Flang] Add parsing and attribute registration for SIMPLE procedures

mlir-maiden committed:
[flang] Extend SIMPLE specifier support: disallow PURE+SIMPLE combination and update purity handling

- Adds a rule in resolve-names.cpp to diagnose procedures declared with both PURE and SIMPLE attributes (Fortran 2023 C759, C1543).
- Updates IsPureProcedureImpl() in tools.cpp so that SIMPLE implies PURE (Fortran 2023 §15.8).
- Introduces a new LIT test (simple-conflict.f90) verifying diagnostic emission for PURE+SIMPLE conflicts.

mlir-maiden commented:
Ping :) - addressed all prior feedback and ready for another look. Thank you

klausler reviewed (commented):
You need tests that verify that cases in Fortran that require a PURE procedure also work with a SIMPLE procedure.

You need to update the intrinsic procedures that used to be PURE and are now documented as being SIMPLE.

mlir-maiden committed:
[flang] Treat SIMPLE procedures as PURE
           - Set PURE implicitly when SIMPLE is specified and add symbol-dump coverage

eugeneepshteyn commented:
@mlir-maiden , please check the following:

(1) It seems that in `CharacterizeProcedure()` you may need to tell procedure being characterized that it's SIMPLE procedure. Consider the following:
```
subroutine test(f)
  interface
    simple subroutine f()
    end subroutine
  interface
end subroutine
```
What are the characteristics of `f()`? Do they include `SIMPLE`?

(2) Please check if the following compiles:
```
module test_simple

abstract interface
  simple subroutine simple_proc()
  end subroutine
end interface

contains

subroutine call_it(p)
  procedure(simple_proc) :: p
end subroutine

simple subroutine my_simple()
end subroutine

subroutine test()
  call call_it(my_simple)
end subroutine
end module
```
(3) The following should work, because ABS (and many other intrinsics) are inherently SIMPLE:
```
simple subroutine test()
  real :: x
  x = abs(-1.0)
end subroutine
```
(4) This should fail:
```
simple impure subroutine bug()
end subroutine
```

mlir-maiden committed:
[flang] Handle SIMPLE propagation and SIMPLE/IMPURE conflict

Propagate SIMPLE in CharacterizeProcedure() so it is reflected in
procedure characteristics.

Treat SIMPLE as satisfying PURE in procedure compatibility, while
preserving that PURE alone does not satisfy SIMPLE.

Diagnose invalid SIMPLE+IMPURE combinations.

Update simple.f90 to use CHECK-DAG since the symbol dump order is not stable.

mlir-maiden committed:
[flang] Fix clang-format in SIMPLE/IMPURE check

eugeneepshteyn commented on flang/lib/Evaluate/tools.cpp:
Please follow `DEREF` pattern in `IsExplicitlyImpureProcedure()`.

Consider the following:
```
module m
  contains
    simple subroutine s()
    end subroutine
end module

program p
  use m
  ! IsSimpleProcedure on the use-assoc symbol for 's' must return true
end program
```
Please double-check this and add a test for this case.

eugeneepshteyn commented on flang/lib/Semantics/resolve-names.cpp:
This and live 700 below doesn't feel right: SetExplicitAttr functions don't seem to be the right place for such exceptions. Please look for other places how such things are handled (e.g., ELEMENTAL is PURE, but PURE is not necessarily ELEMENTAL).

mlir-maiden replied:
Done

eugeneepshteyn commented on flang/lib/Evaluate/tools.cpp:
We now use `F2023 15.8` spec reference format.

mlir-maiden replied:
Done

eugeneepshteyn commented on flang/lib/Semantics/check-declarations.cpp:
Please add a comment with the spec reference `F2023 C1553`.

eugeneepshteyn reviewed (changes requested):
I reviewed the whole PR again.

Also, we should consider modifying `flang/lib/Evaluate/intrinsics.cpp` to add SIMPLE there. Or was this planned as a separate PR?

eugeneepshteyn commented:
Please also rebase this to the latest LLVM.

mlir-maiden commented:
Thanks for the review, Eugene.
This PR seems to have grown beyond its original scope (parsing + attribute registration) into some semantic handling (SIMPLE->PURE, implicit/explicit bookkeeping, semantics conflict diagnostics, tests).
Should I:
narrow it back to parsing/attr registration and move semantics to follow-ups, or
continue expanding this PR to cover SIMPLE semantics?
Just wanted to highlight this and get your guidance on the preferred direction.

eugeneepshteyn commented:
> Should I: narrow it back to parsing/attr registration and move semantics to follow-ups, or continue expanding this PR to cover SIMPLE semantics? 

I think it would be worth expanding this PR to cover SIMPLE semantics.

eugeneepshteyn commented:
> Also, we should consider modifying `flang/lib/Evaluate/intrinsics.cpp` to add SIMPLE there. Or was this planned as a separate PR?

Let's do this as a separate PR. Please leave a TODO comment in that file mentioning this.

@mlir-maiden , could you please also modify SIMPLE section in this document with the status of current implementation:

https://github.com/llvm/llvm-project/blob/main/flang/docs/F202X.md

You don't have to remove any existing text there, but please add "current status" section so that people know that the works is being done and what may not be complete.

Besides the intrinsics, let's have semantic checks done as much as possible and add TODO for any obviously missing pieces.

mlir-maiden commented:
> > Also, we should consider modifying `flang/lib/Evaluate/intrinsics.cpp` to add SIMPLE there. Or was this planned as a separate PR?
> 
> Let's do this as a separate PR. Please leave a TODO comment in that file mentioning this.
> 
> @mlir-maiden , could you please also modify SIMPLE section in this document with the status of current implementation:
> 
> https://github.com/llvm/llvm-project/blob/main/flang/docs/F202X.md
> 
> You don't have to remove any existing text there, but please add "current status" section so that people know that the works is being done and what may not be complete.
> 
> Besides the intrinsics, let's have semantic checks done as much as possible and add TODO for any obviously missing pieces.

Understood, Eugene. I added TODO in intrinsics.cpp and updated F202X.md with a "Current status" section.

For this PR, I covered the following support  and semantics for SIMPLE:
- Parsing and symbol attribute registration for SIMPLE procedures
- SIMPLE is represented in procedure characteristics
- SIMPLE procedures are accepted where PURE procedures are required
- PURE procedures do not satisfy SIMPLE requirements
- SIMPLE/IMPURE conflicts are diagnosed
- Use-associated SIMPLE procedures are recognized (DEREF/GetUltimate handling)
- Added tests for all of the above
I plan to handle intrinsic procedures being marked as SIMPLE in a follow-up PR.

mlir-maiden committed:
[flang] Address SIMPLE procedure review feedback

- Move SIMPLE=>PURE handling from SetExplicitAttr* to IsPureProcedureImpl
- Derive PURE via query (not stored as a symbol attribute)
- Update IsSimpleProcedure to use GetUltimate/GetMainEntry
- Fix SIMPLE detection for ENTRY and use-associated procedures
- Add use-associated SIMPLE regression test
- Remove outdated symbol-dump test (simple.f90)
- Add current-status section for SIMPLE in F202X.md
- Add TODO for marking intrinsics as SIMPLE in intrinsics.cpp (follow-up PR)

kparzysz commented on flang/test/Semantics/simple-satisfies-pure-negative.f90:
```diff
@@ -0,0 +1,36 @@
+! This is the negative/control case for simple-satisfies-pure.f90.
+! It verifies that a procedure which is neither PURE nor SIMPLE is rejected
+! when passed to a dummy argument that requires a PURE procedure.
+! RUN: not %flang_fc1 -fsyntax-only %s 2>&1 | FileCheck %s
```
Please use the test_errors.py script for checking diagnostics.

mlir-maiden replied:
Updated to use test_errors.py.

mlir-maiden replied:
Should I also update `flang/test/Semantics/pure-does-not-satisfy-simple.f90` and `flang/test/Semantics/simple-impure-conflict.f90` to use `test_errors.py`, or keep the change limited to this test? Thank you

kparzysz replied:
Please use it every time you're checking that a specific diagnostic message has been emitted.  That applies to these two files you mentioned as well.

mlir-maiden committed:
[flang] Use test_errors.py for SIMPLE negative diagnostic test

akuhlens commented on flang/docs/F202X.md:
```diff
@@ -198,6 +209,7 @@ a powerful sum data type as in Haskell or Rust.  Unlike the
 current `ENUM, BIND(C)` feature, `ENUMERATION TYPE` defines a new
 type name and its distinct values.
 
+
```
Please remove miscellaneous whitespace changes.

akuhlens reviewed (commented):
This is looking good to me. I think the only thing I would like to see is that you merge all the semantic checks into a single test or at least consolidate as possible. Note, `test_erros.py` checks that the error messages that are specified are present and that there are no other error messages. So at the very least the two `syntax-only` tests can be combined into the negative test.

mlir-maiden commented:
> This is looking good to me. I think the only thing I would like to see is that you merge all the semantic checks into a single test or at least consolidate as possible. Note, `test_erros.py` checks that the error messages that are specified are present and that there are no other error messages. So at the very least the two `syntax-only` tests can be combined into the negative test.

Roger that. Thank you

eugeneepshteyn commented:
It seems that need to add SIMPLE to `subprogramPrefixAttrs` in Semantics/mod-file.cpp. 

Please also add a test that SIMPLE can be serialized to a module file and could be read from it. See flang/test/Semantics/modfile04.f90 for an example.

eugeneepshteyn commented:
Besides the module file stuff, I think this is really close!

mlir-maiden committed:
[flang] Add SIMPLE to mod files and consolidate semantic tests

- Add SIMPLE to subprogramPrefixAttrs in mod-file.cpp so it is
  serialized to and read from module files
- Add modfile test to verify SIMPLE is preserved in .mod files
- Consolidate overlapping semantic tests:
  * SIMPLE + IMPURE conflict
  * PURE does not satisfy SIMPLE
  * Neither PURE nor SIMPLE satisfies PURE
  into a single test file
- Remove redundant tests replaced by the consolidated coverage

mlir-maiden committed:
[flang] Apply clang-format to mod-file.cpp

eugeneepshteyn commented on flang/test/Semantics/modfile-simple.f90:
```diff
@@ -0,0 +1,93 @@
+! Module with a subroutine with alternate returns
+module m3
+contains
+  subroutine altReturn(arg1, arg2, *, *)
```
Was this supposed to be `simple`?

eugeneepshteyn commented on flang/test/Semantics/modfile-simple.f90:
```diff
@@ -0,0 +1,93 @@
+  end subroutine
+end module m3
+
+!Expect: m1.mod
```
If test tool allows it, could you please try to move the `!Expect` sections to follow right after the source? It makes it easier to review.

eugeneepshteyn commented on flang/test/Parser/simple.f90:
```diff
@@ -0,0 +1,9 @@
+! RUN: %flang_fc1 -fdebug-dump-parse-tree %s | FileCheck %s
+
```
Nit: please remove the empty line

eugeneepshteyn commented on flang/test/Semantics/simple-satisfies-pure.f90:
```diff
@@ -0,0 +1,33 @@
+! This test verifies that SIMPLE procedures satisfy PURE requirements.
```
Additional tests to add for better coverage (add to any test file where appropriate):

1. `simple integer function f(x)` - SIMPLE with function that has a type spec.
2. I think `simple pure` and `pure simple` should be allowed. Please add a test for that.
3. Similarly, where you test `impure simple` also test `simple impure`.
4. Test `module simple function` (let me know if you need an example for this)
5. A `simple` function that returns some kind of non-trivial `character` type.

mlir-maiden replied:
Roger that. I will add test coverage and will consolidate these two as well:
`simple-satisfies-pure.f90`
`simple-use-associated.f90`

eugeneepshteyn reviewed (commented):
Thank you, moving in the right direction! I concentrated on reviewing the test coverage this time.

mlir-maiden committed:
[flang] Extend SIMPLE attribute test coverage

- Consolidate simple-satisfies-pure.f90 and simple-use-associated.f90
  into simple-positive.f90
- Add coverage for:
  * simple pure / pure simple
  * simple impure / impure simple conflicts
  * SIMPLE functions returning non-trivial CHARACTER results
  * SIMPLE separate module procedures (module/simple order)
- Update modfile tests

eugeneepshteyn reviewed (approved):
LGTM!

@mlir-maiden , thank you for staying with it! :)

I'm going to run some tests and I can submit it after that.

mlir-maiden commented:
> LGTM!
> 
> @mlir-maiden , thank you for staying with it! :)
>
Thank you Eugene and team. This has definitely been a team effort. I’ll follow up next with updating the intrinsic procedures that were previously treated as PURE to reflect their SIMPLE designation per F2023. Please let me know if there are any additional areas you would like me to cover as part of this work.

eugeneepshteyn committed:
Merge branch 'main' into simple-clean

eugeneepshteyn commented:
@mlir-maiden , could you please update PR title and description to what you'd like to see in the git commit message? I think the info there is a bit out of date with the current implementation. I'm ready to submit it otherwise.

mlir-maiden commented:
Yes, in the next 1 hour
